### DISCUSSION TITLE
51515, Inverted pendulum robot, running on MPC

### DISCUSSION THREAD

**Discussion by tobiasskov** [pybricks/support#948] [Q&A] [answered]
Hi Pybricks team and community

First of all – thank you for developing such awesome tools for getting the most out of our Lego hubs.

I am trying to code an inverted pendulum robot using the 51515 set, running on Model Predictive Control. As a first step I am doing system identification for a [DC motor model](https://ctms.engin.umich.edu/CTMS/index.php?example=MotorSpeed&section=SystemModeling) and have some questions that I hope you can help with:

1. How is Pulse-Width-Modulation (PWM) implemented in Pybricks? I.e. what is the relationship between the duty cycle input for the motor and the voltage applied across the motor? 
2. I have been running some step response test where a single motor is connected to the hub. The test is running the single motor at duty cycle set to 0%, 30%, 100% each for 3 s. To compare between the 4 motors I have been running the program on all 4 motors and collected all the output into an interactive plotly figure in the attached html [figures.zip](https://github.com/pybricks/support/files/10715658/figures.zip). As you can see from the highlighted pictures (one.png, two.png) at random position the sampling frequency changes. The motors should however be able to provide measurements at 100 Hz consistently [Motor technical sheet](https://le-www-live-s.legocdn.com/sc/media/files/support/spike-prime/techspecs_technicmediumangularmotor-19684ffc443792280359ef217512a1d1.pdf). I found a similar thing happening when logging hub voltage and current with no motors attached (hub_only.png)
Any ideas why this is happening? And maybe how to get around it?

All pybricks code and post-processing code can be found in this [github repo](https://github.com/TobySkov/51515-System-Identification).

I am aware that running MPC using pybricks, would require extending the Pybricks firmware with additional matrix routines such as LU-factorization and I am very interested in contributing with this.  

Lastly I will leave you with some projects that serves as inspiration:
[Inverted pendulum robot](https://www.youtube.com/watch?v=7S66NxwkAxU)
[Two legged wheeled robot](https://www.youtube.com/watch?v=07z1FbjhixM)

Thank you, Tobias

**Comment by laurensvalk** [comment, marked as answer]
> the sampling frequency changes

Can you highlight the piece of your code where you sample the values? Occasionally the script will take some time to garbage-collect unused data, but 80 ms does seem a little long.

Slightly off topic --- this brings back memories of https://github.com/ev3dev/ev3dev/issues/324. I was more or less in the same situation. So maybe you'll end up building something like Pybricks in 8 years from now :smile:

**Reply by tobiasskov** [reply]
Thank you for the quick reply!

The pybricks code used for the test can be found here: [Github](https://github.com/TobySkov/51515-System-Identification/blob/main/pybricks-backup-2023-02-11_11-33-55/single_motor_test.py).
Also an example of the generated output can be found here: [Github](https://github.com/TobySkov/51515-System-Identification/blob/main/test_results/motor_1_program_1.txt). The output was copy-pasted from the terminal and the 51515 hub was connected to the online pybricks IDE using bluetooth. 

> Slightly off topic --- this brings back memories of [ev3dev/ev3dev#324](https://github.com/ev3dev/ev3dev/issues/324). I was more or less in the same situation. So maybe you'll end up building something like Pybricks in 8 years from now 😄

Very interesting read, I will have to study it further to understand all the concepts being discussed. I don't think I would be able to develop something like pybricks from scratch and would rather contribute to the pybricks project in any meaningfull way 😄  This also got me thinking, do you deem it realistic to develop a MPC controlled robot on the 51515 hardware (constraints)? Maybe the it is necesseary to add a rasberry pi in the mix for running the optimization algorithm, kalman filter, etc.

**Reply by laurensvalk** [reply]
Printing values over Bluetooth can take a non deterministic amount of time.

You could consider keeping the data in memory and print it all after the experiment is complete.

Most of the information in the referenced issue may not be relevant. That was about a Linux system while the inventor hub runs pybricks without an operating system underneath. It was more the overal similarity of the challenges and project type that brought back some memories :)

It is possible to implementation portions of your solution in C if needed, so you could use this if computational complexity becomes an issue.

**Reply by tobiasskov** [reply]
Great, thank you - this was the cause of the changes in sampling frequency.
I have updated the pybricks code, test data and test figures accordingly at the [github repo](https://github.com/TobySkov/51515-System-Identification). It is interesting to see how the four motors in my 51515 set have different angular speeds during steady state despite the same duty cycle input and similar currents drawn.

Do you have any suggestions for a conversion between duty cycle and applied voltage (pwm)?

**Reply by dlech** [reply]
It's not easy to get an accurate value for this since the only available measured values are the overall battery voltage and current. Of course, just using the battery voltage times the duty cycle will get you in the ballpark. 

If the motor is always under low load, you could probably subtract 0.2V from the measured battery voltage to get a bit more accurate (this number is an estimate from some experiments I did a few years ago where I measured a 0.15V drop between the battery and the motor when the motor was run at 100% duty cycle with no load).

**Reply by tobiasskov** [reply]
Thank you - I will try running some tests using this assumption :)
